Added overview of OTEL (#2964)
spryker/spryker-docs: docs/dg/dev/backend-development/opentelemetry/overview.md

{
    /**
     * @return array<\Spryker\Service\MonitoringExtension\Dependency\Plugin\MonitoringExtensionPluginInterface>
     */
    protected function getMonitoringExtensions(): array
    {
        return [
            new OpentelemetryMonitoringExtensionPlugin(),
        ];
    }
}

```
After that you can call method from Monitoring service and they will be transalted to Opentelemtry action. Be advised that some of the methods are not covered due to the fact  that those things have no direct implementation. E.g. `\Spryker\Service\Opentelemetry\Plugin\OpentelemetryMonitoringExtensionPlugin::markAsConsoleCommand` does nothing and transaction is marked as console command automatically.

### Wire a console command

```php
<?php

namespace Pyz\Zed\Console;

...
use Spryker\Zed\Opentelemetry\Communication\Plugin\Console\OpentelemetryGeneratorConsole;
...

class ConsoleDependencyProvider extends SprykerConsoleDependencyProvider
{
    protected function getConsoleCommands(Container $container): array
    {
        $commands = [
            ...
            new OpentelemetryGeneratorConsole(),
            ...
        ];

        return $commands;
    }
}

```

> Wire new console command into your install script. It allows it to be executed on every deploy and generated filed during it.

Reviewer: Please check this part `and generated filed during it.`

`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getExcludedSpans` - allows you to exclude specific spans that you noticed in your traces. You need to use specific span name for that.
`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getPathPatterns` - allows to add a path pattern that shows where console command should look for you classes to cover them with hooks. OOTB all Spryker directories are covered together with `Pyz` namespace on the project. Be advised that you should not cover code that is autogenerated.
`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getOutputDir` - specifies in what directory you want to put generated hooks. Default value is `src/Generated/OpenTelemetry/Hooks/`.
`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::areOnlyPublicMethodsInstrumented` - OOTB we cover with hooks only public methods (except Controller classes), but you can change that with this configuration method.

> `\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getCriticalClassNamePatterns` - some of the spans can be marked as critical in order to give them more priority on sampling. You can define what classes you are interested in for that. OOTB we have Controllers and Facades here.

Reviewer: This is core level classes, i would provide example how it can be changed.

> ---

Reviewer 2: We need to add data about New Relic + OTEL

Author: What do you mean?

Reviewer 3: `It's impossible to have NewRelic and OTel enabled on one environment.`

> title: Opentelemetry overview

Reviewer 3: ```suggestion
title: OpenTelemetry overview
```
I see different writings in the document, please align.

                    }

                    $span->setStatus($exception !== null ? \OpenTelemetry\API\Trace\StatusCode::STATUS_ERROR : \OpenTelemetry\API\Trace\StatusCode::STATUS_OK);

                    $span->end();
                }
            );
```

### Collector
Collector does what you think it is. It collects traces and send it to your backend after. Sending traces to collector is done after that request is sent, so it will not make a client wait just for that.

## Integration
For the integration you can use a script from a [Installer repo](https://github.com/spryker/opentelemetry-installer/). Use the latest version and run it in your system.
Below you can see what steps this script executes if you want to do everything manually.

### Install required packages
OpenTelementry provides instrumentation via packages that can be installed and register hooks automatically. If you want to instrument something that is not covered in our code base, but is required for you, you can try to install packages that are listed [here](https://opentelemetry.io/ecosystem/registry/?language=php) or something that you consider useful for you from other sources.

> Spryker provides [spryker/opentelemetry](https://packagist.org/packages/spryker/opentelemetry) package that covers essential parts. It includes the entry point for instrumentation, plugin that you can wire in your monitoring service and a console command that allows you to generate [hooks](https://opentelemetry.io/docs/zero-code/php/#how-it-works) that will generate spans automatically. Also it includes instrumentation of Propel, Redis, ElasticSeach and RabbitMQ calls.

Reviewer 3: ```suggestion
Spryker provides [spryker/opentelemetry](https://packagist.org/packages/spryker/opentelemetry) package that covers essential parts. It includes the entry point for instrumentation, plugin that you can wire in your monitoring service and a console command that allows you to generate [hooks](https://opentelemetry.io/docs/zero-code/php/#how-it-works) for project's code, which will create spans automatically. Also it includes instrumentation of Propel, Redis, ElasticSearch and RabbitMQ calls.
```

In addition, you may want to install a `open-telemetry/opentelemetry-auto-guzzle` package to cover Guzzle calls to 3rd party services or to the services in your application.

### (Optional) Install Monitoring module
Opentelemetry integration doesn't require to use Monitoring service, but this is highly recommended as it allows you to add custom attributes, change your traces (transaction) name during the request execution and have exceptions added to the root span for visibility.
You can get a module [here](https://packagist.org/packages/spryker/monitoring).

> After install you can wire a Monitoring plugin from a `spryker/opentelemtry` module to get all listed features.

Reviewer 3: ```suggestion
After installation you can wire a Monitoring plugin from a `spryker/opentelemtry` module to enable all listed above features.
```

```php
<?php

namespace Pyz\Service\Monitoring;

use Spryker\Service\Monitoring\MonitoringDependencyProvider as SprykerMonitoringDependencyProvider;
use Spryker\Service\Opentelemetry\Plugin\OpentelemetryMonitoringExtensionPlugin;

class MonitoringDependencyProvider extends SprykerMonitoringDependencyProvider
{
    /**
     * @return array<\Spryker\Service\MonitoringExtension\Dependency\Plugin\MonitoringExtensionPluginInterface>
     */
    protected function getMonitoringExtensions(): array
    {
        return [
            new OpentelemetryMonitoringExtensionPlugin(),
        ];
    }
}

```

> After that you can call method from Monitoring service and they will be translated to OpenTelemetry action. Be advised that some of the methods are not covered due to the fact that those things have no direct implementation, like `\Spryker\Service\Opentelemetry\Plugin\OpentelemetryMonitoringExtensionPlugin::markStartTransaction()` as transaction will start anyway.

Reviewer 3: ```suggestion
After that you can call methods from Monitoring service and they will be translated to OpenTelemetry action. Be advised that some of the methods are empty due to the fact that those things have no direct implementation in Opentelemetry, like `\Spryker\Service\Opentelemetry\Plugin\OpentelemetryMonitoringExtensionPlugin::markStartTransaction()` as transaction will start anyway.
```

### Wire a console command

> Provided console command used to generate hooks automatically for classes that you want to be covered with spans.

Reviewer 3: ```suggestion
Console command *OpentelemetryGeneratorConsole*  is used to generate hooks automatically for classes that you want to be covered with spans.
```

```php
<?php

namespace Pyz\Zed\Console;

...
use Spryker\Zed\Opentelemetry\Communication\Plugin\Console\OpentelemetryGeneratorConsole;
...

class ConsoleDependencyProvider extends SprykerConsoleDependencyProvider
{
    protected function getConsoleCommands(Container $container): array
    {
        $commands = [
            ...
            new OpentelemetryGeneratorConsole(),
            ...
        ];

        return $commands;
    }
}

```

> Wire new console command into your install script. It allows it to be executed on every deploy and generated filed during it.

Reviewer 3: ```suggestion
You have to wire this console command into your install script, so it's executed on every deployment.
Please make sure that this command is located after all code modifications and generation.
```

            ...
        ];
    }

}
```

`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getPathPatterns()` - allows to add a path pattern that shows where console command should look for you classes to cover them with hooks. OOTB all Spryker directories are covered together with `Pyz` namespace on the project. Be advised that you should not cover code that is autogenerated.
```php
class OpentelemetryConfig extends AbstractBundleConfig
{
    public function getPathPatterns(): array
    {
        return [
            '#^vendor/spryker/[^/]+/.*/.*/(Zed|Shared)/.*/(?!Persistence|Presentation)[^/]+/.*#',
            '#^vendor/spryker/[^/]+/Glue.*#',
            '#^vendor/spryker(?:/spryker)?-shop/[^/]+/.*#',
            '#^vendor/spryker-eco/[^/]+/.*#',
            '#^src/Pyz/.*#',
        ];
    }

}
```

`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getOutputDir()` - specifies in what directory you want to put generated hooks. Default value is `src/Generated/OpenTelemetry/Hooks/`.
```php
class OpentelemetryConfig extends AbstractBundleConfig
{
    public function getOutputDir(): string
    {
        return APPLICATION_SOURCE_DIR . '/Generated/OpenTelemetry/Hooks/';
    }

}
```

> `\Spryker\Zed\Opentelemetry\OpentelemetryConfig::areOnlyPublicMethodsInstrumented()` - OOTB we cover with hooks only public methods (except Controller classes), but you can change that with this configuration method.

Reviewer 3: ```suggestion
`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::areOnlyPublicMethodsInstrumented()` - OOTB we cover with hooks  public methods of all classes and all methods of Controller classes, but you can change that with this configuration method.
```

btw, I'm not sure that private methods could be hooked in.

```php
class OpentelemetryConfig extends AbstractBundleConfig
{
    public function areOnlyPublicMethodsInstrumented(): bool
    {
        return true;
    }

}
```

`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getCriticalClassNamePatterns()` - some of the spans can be marked as critical in order to give them more priority on sampling. You can define what classes you are interested in for that. OOTB we have Controllers and Facades here.
```php
class OpentelemetryConfig extends AbstractBundleConfig
{

>     public function getCriticalClassNamePatterns(): array

Author: updated description, but the answer is no. Currently those are just strings, but you are right. We will update the method that is using it eventually to use also patterns

    {
        return [
            'Facade',
            'Controller',
        ];
    }

}
```

### Enable PHP extension

> Hooks processing require you to have a few PHP extensions in place. Spryker prepares our PHP images, so you need to install nothing, but just enable it in your deploy file

Reviewer 3: ```suggestion
Hooks processing requires you to have a few PHP extensions in place. Spryker has prepared a new PHP image, so you need to install nothing, but just enable it in your deploy file
```

```yaml
namespace: spryker-otel
tag: 'dev'

environment: docker.dev
image:

>     tag: spryker/php:8.3

Reviewer 3: ```suggestion
    tag: spryker/php:8.3-alpine3.20-otel
```

    php:
        enabled-extensions:
            - opentelemetry
            - grpc
            - protobuf
```

> Be advised that `blackfire` extension can have a conflict with a `opentelemetry` one, so make sure that you are not using both in the same time.

Reviewer 3: what about NewRelic?

### Span criticality
Some of the spans are more relevant for the traces, some of them are not. So for span sampling we have 3 different span criticality: non critical, regular and critical one.
They are using different probability and execution time limits that can be configured separately.

#### Critical spans
All RabbitMQ and ElasticSearch spans are marked as critical OOTB.
Propel INSERT/DELETE/UPDATE calls also considered as critical.
Hooks for classes that are mentioned in `\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getCriticalClassNamePatterns()` method also considered as critical (OOTB Facades and Controllers)
#### Non Critical spans
OOTB all SELECT queries are considered as non critical.
#### Regular spans
All the other generated hooks are regular ones.

#### How to make span critical or not.
By setting a span attribute `no_critical` and `is_critical`, span can be marked as critical or not.

### Sampling configuration
Like already mentioned, valued that used for sampling can be changed. In order to do so, you need to change a few env variables.

> `OTEL_BSP_MIN_SPAN_DURATION_THRESHOLD` - this is value that is used in `Closing span sampling` for critical and regular spans. OOTB the value is `20`. This value will be used as nanosecond.

Reviewer 3: I don't see a natural language description of this parameter.
```suggestion
`OTEL_BSP_MIN_SPAN_DURATION_THRESHOLD` - this is value that is used in `Closing span sampling` for critical and regular spans. OOTB the value is `20`. This value will be used as nanosecond.
```

`OTEL_BSP_MIN_CRITICAL_SPAN_DURATION_THRESHOLD` - this is similar value to `OTEL_BSP_MIN_SPAN_DURATION_THRESHOLD` one, but used for non critical spans. Default value is `10`.

> `OTEL_TRACES_SAMPLER_ARG` - this is a probability for `Opening span sampling` and is used for the regular spans. Default value is `0.3`.

Author: tried to add more explanation to sampling "chapter"

### Additional configuration
`OTEL_SERVICE_NAMESPACE` - defines a namespace that is shown in the resource attributes of trace. Default value is `spryker`.
`OTEL_EXPORTER_OTLP_ENDPOINT` - defines a collector endpoint. Should be configured for your project as it varies from setup to setup. Default value is `http://collector:4317`.
`OTEL_SERVICE_NAME_MAPPING` - a json value that can be provided to define a service name is `MonitoringService::setApplicationName()` is not useful for you. By default OOTB Spryker application are mapped. `\Spryker\Service\Opentelemetry\OpentelemetryInstrumentationConfig::getServiceNameMapping()` can be checked for example.
`OTEL_DEFAULT_SERVICE_NAME` - if no mapping is provided and no name is provided via `MonitoringService::setApplicationName()` this value would be used as a fallback value. Default value is `Default Service`.
`OTEL_BSP_SCHEDULE_DELAY` - this value is used in SpanProcessor in order to flush all processed spans into exported after some delay. Default value is `1000` milliseconds.
`OTEL_BSP_MAX_QUEUE_SIZE` - defines max amount of spans that should be processed. All spans on top will be dropped. Default value is `2048`

> `OTEL_BSP_MAX_EXPORT_BATCH_SIZE` - define a spans batch size that should be sent to an exporter. Default value `512`.

Author: Exporter is a part of application and collector is a thing that you have in your own container that exporter send data to

## Recommendations

Due to the fact that we use PHP code to instrument codebase, you should consider performance. Tracing is an expensive operation and we should use it wisely.

Please minimise amount of generated spans per request. OpenTelemetry documentation recommends to have no more than 1000 of them. So you can skip some spans via configuration that are not relevant to you. Don't be afraid, errors will be processed even if the method was not instrumented as Error Event will be attached to the root span.

Use sampling to not get a full trace every time. Please check configuration section for the reference.

> Skip some traces. You may not want to get a full trace for all of your transactions. You can define a probability of detailed trace overview by setting a probability via `OTEL_TRACE_PROBABILITY` env variable. Be advised that Trace still will be processed and root span will be there for you. Also requests that are changing something in your application (POST, DELETE, PUT, PATCH) considered as critical and will be processed anyway.

Reviewer 3: I miss explanation of the NON-INSTRUMENTED classes, that are restricted in the main Spryker module.
For example, Container calls, - \Spryker\Zed\Opentelemetry\OpentelemetryConfig::getExcludedDirs
I already faced issues that some spans are missing.

Author: added more explanation for those methods

> Skip some traces. You may not want to get a full trace for all of your transactions. You can define a probability of detailed trace overview by setting a probability via `OTEL_TRACE_PROBABILITY` env variable. Be advised that Trace still will be processed and root span will be there for you. Also requests that are changing something in your application (POST, DELETE, PUT, PATCH) considered as critical and will be processed anyway.

Reviewer 3: Please include the possible values for each setting, i.e. 
OTEL_TRACE_PROBABILITY = integer 1...100

Author: updated

---
title: OpenTelemetry overview
description: Describing what is OpenTelemetry and how Spryker integrates it in the SCCOS
template: howto-guide-template
last_updated: Dec 5, 2024
related:
---

> Spryker allows to integrate different monitoring tools like NewRelic into the SCCOS. But usually those tools are vendor locked and allows to work only with limited amount of backends or have not so much customizability options.

Reviewer 3: I don't know where, but let's add a mention, similar to the following
```
OTel is a paid service and requires additional contract with Spryker, please talk to our Sales Team.
```

After installation you can wire a Monitoring plugin from a `spryker/opentelemtry` module to enable all listed above features.
```php
<?php

namespace Pyz\Service\Monitoring;

use Spryker\Service\Monitoring\MonitoringDependencyProvider as SprykerMonitoringDependencyProvider;
use Spryker\Service\Opentelemetry\Plugin\OpentelemetryMonitoringExtensionPlugin;

class MonitoringDependencyProvider extends SprykerMonitoringDependencyProvider
{
    /**
     * @return array<\Spryker\Service\MonitoringExtension\Dependency\Plugin\MonitoringExtensionPluginInterface>
     */
    protected function getMonitoringExtensions(): array
    {
        return [
            new OpentelemetryMonitoringExtensionPlugin(),
        ];
    }
}

```
After that you can call methods from Monitoring service and they will be translated to OpenTelemetry actions. Be advised that some of the methods are empty due to the fact that those things have no direct implementation in Opentelemetry, like `\Spryker\Service\Opentelemetry\Plugin\OpentelemetryMonitoringExtensionPlugin::markStartTransaction()` as transaction will start anyway.

### Wire a console command

> Spryker is a big application, so you definitely doesn't want to prepare all hooks manually. Console command *OpentelemetryGeneratorConsole*  is used to generate those hooks automatically for classes that you want to be covered with spans.

Reviewer 3: ```suggestion
Spryker is a big application, so you definitely don't want to create all hooks manually. Console command  class *OpentelemetryGeneratorConsole*  is used to generate those hooks automatically for classes that you want to cover with spans.
```

```php
<?php

namespace Pyz\Zed\Console;

...
use Spryker\Zed\Opentelemetry\Communication\Plugin\Console\OpentelemetryGeneratorConsole;
...

class ConsoleDependencyProvider extends SprykerConsoleDependencyProvider
{
    protected function getConsoleCommands(Container $container): array
    {
        $commands = [
            ...
            new OpentelemetryGeneratorConsole(),
            ...
        ];

        return $commands;
    }
}

```

> You have to wire this console command into your install script, so it's executed on every deployment.

Reviewer 3: ```suggestion
You have to wire this console command into your install script, so it's executed on every deployment during base container image build. We recommend placing it as the last one in the section **build**.
```

{
    public function getOutputDir(): string
    {
        return APPLICATION_SOURCE_DIR . '/Generated/OpenTelemetry/Hooks/';
    }

}
```

`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::areOnlyPublicMethodsInstrumented()` - OOTB we cover with hooks only public methods of regular classes and all methods of Controller classes, but you can change that with this configuration method.
```php
class OpentelemetryConfig extends AbstractBundleConfig
{
    public function areOnlyPublicMethodsInstrumented(): bool
    {
        return true;
    }

}
```

`\Spryker\Zed\Opentelemetry\OpentelemetryConfig::getCriticalClassNamePatterns()` - some of the spans can be marked as `critical` in order to give them more priority on sampling. You can define what classes you are interested in for that. OOTB we have Controllers and Facades here. With the current implementation it's not a Regex, but just a string that we are looking for in the class name.
```php
class OpentelemetryConfig extends AbstractBundleConfig
{
    public function getCriticalClassNamePatterns(): array
    {
        return [
            'Facade',
            'Controller',
        ];
    }

}
```

> 

Reviewer 3: ```suggestion

Please get yourself familiar with the predefined configuration of this module in the vendor folder. Based on our research, some Core classes are excluded from instrumentation.
```

> ### Enable PHP extension

Reviewer 3: ```suggestion
### Enable PHP extensions
```

> Hooks processing requires you to have a few PHP extensions in place. Spryker has prepared a new PHP image, so you need to install nothing, but just enable it in your deploy file.

Reviewer 3: ```suggestion
Hooks processing requires you to have a few PHP extensions in place. Spryker has prepared a new PHP image, so you need to install nothing, just enable them in your deploy file.
```

```yaml
namespace: spryker-otel
tag: 'dev'

environment: docker.dev
image:
    tag: spryker/php:8.3-alpine3.20-otel
    php:
        enabled-extensions:
            - opentelemetry
            - grpc
            - protobuf
```

> Be advised that `blackfire` extension can have a conflict with a `opentelemetry` one, so make sure that you are not using both in the same time. We also would not recommend to run it together with a `newrelic` one.

Reviewer 3: Let's be more helpful here.
Blackfire can have conflicts - how is this visible if still both are enabled?
Is there an option how to add a switcher?

We would recommend to do so using `MonitoringService::addCustomParameter()` method. By default Spryker provides a few of those attributes already, like a current store name and locale.
If you don't want to use `MonitoringService` for some reason, you may use `\Spryker\Service\Opentelemetry\OpentelemetryService::setCustomParameter` directly. It does exactly the same job, but without necessity to install `Monitoring` module.
All attributes added via those services will be added into the root span on the very end of request execution, so you can call it even after response wes already sent.

## Custom events

If you want to add some custom event to you trace in order to configure some logic into your backend, you can trigger a custom event during the execution. It works pretty much the same as adding custom attributes, so it's pretty easy to use.
The same as for custom attributes you can add those via `\Spryker\Service\Monitoring\MonitoringService::addCustomEvent()` or via `\Spryker\Service\Opentelemetry\OpentelemetryService::addEvent()` if you dont want to install `Monitoring` module.
Event will be wired into the root span.

## Error handling

> You are definitely want to track all possible errors during the execution. OOTB OpenTelemetry integration will catch all exceptions that were thrown during span execution. But there are possibility that error can be lost if part of your code was not covered with hooks, or spans were sampled out, or trace was marked as not detailed.

Reviewer 3: ```suggestion
You are definitely want to track all possible errors during the execution. OOTB OpenTelemetry integration will catch all exceptions that were thrown during span execution. But there is a possibility that error is lost if part of your code was not covered with hooks, or spans were sampled out, or trace was marked as not detailed.
```

Reviewer 3: @gechetspr is it possible to have a configuration that makes sure that ALL exceptions are attached to the root span, no matter what.
